transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.util.internal.PlatformDependent;
+
+final class IOUringCompletionQueue {
+
+  //these offsets are used to access specific properties
+  //CQE (https://github.com/axboe/liburing/blob/master/src/include/liburing/io_uring.h#L162)
+  private static final int CQE_USER_DATA_FIELD = 0;
+  private static final int CQE_RES_FIELD = 8;
+  private static final int CQE_FLAGS_FIELD = 12;
+
+  private static final int CQE_SIZE = 16;
+
+  private static final int IORING_ENTER_GETEVENTS = 1;
+
+  //these unsigned integer pointers(shared with the kernel) will be changed by the kernel
+  private final long kHeadAddress;
+  private final long kTailAddress;
+  private final long kringMaskAddress;
+  private final long kringEntries;
+  private final long kOverflowAddress;
+
+  private final long completionQueueArrayAddress;
+
+  private final int ringSize;
+  private final long ringAddress;
+  private final int ringFd;
+
+  IOUringCompletionQueue(long kHeadAddress, long kTailAddress, long kringMaskAddress, long kringEntries,
+      long kOverflowAddress, long completionQueueArrayAddress, int ringSize, long ringAddress, int ringFd) {
+    this.kHeadAddress = kHeadAddress;
+    this.kTailAddress = kTailAddress;
+    this.kringMaskAddress = kringMaskAddress;
+    this.kringEntries = kringEntries;
+    this.kOverflowAddress = kOverflowAddress;
+    this.completionQueueArrayAddress = completionQueueArrayAddress;
+    this.ringSize = ringSize;
+    this.ringAddress = ringAddress;
+    this.ringFd = ringFd;
+  }
+
+  public IOUringCqe peek() {
+    long head = toUnsignedLong(PlatformDependent.getIntVolatile(kHeadAddress));
+
+    if (head != toUnsignedLong(PlatformDependent.getInt(kTailAddress))) {
+        long index = head & toUnsignedLong(PlatformDependent.getInt(kringMaskAddress));
+        long cqe = index * CQE_SIZE + completionQueueArrayAddress;
+
+        long eventId = PlatformDependent.getLong(cqe + CQE_USER_DATA_FIELD);
+        int res = PlatformDependent.getInt(cqe + CQE_RES_FIELD);
+        long flags = toUnsignedLong(PlatformDependent.getInt(cqe + CQE_FLAGS_FIELD));
+
+        //Ensure that the kernel only sees the new value of the head index after the CQEs have been read.
+        PlatformDependent.putIntOrdered(kHeadAddress, (int) (head + 1));
+
+        return new IOUringCqe(eventId, res, flags);
```

Comment:
I see 2 problems here (I'm sure you see them as well)

1: litter
2: data copying

The litter part can easily be resolved; but the data copying part is a bit harder.

Reply:
yeah I'm aware of this litter problem, the same for other classes like `Event`
to expose the index into CQE-array is actually a good idea 👍 
I don't think I understand what you mean by 'copying data problem', why is the copying of primitive data types a performance issue? we would still copying data in utility methods right?

Reply:
If you would use an intermediate Java object to get data from the completion queue entries which to the 'caller'. 

Then you go from c record -> java object -> caller. 

So you have an intermediate copy of the data and this is one of the things io_uring tries to prevent if I understand the design correctly. Per entry, this would be 64 bytes of wasted memory bandwidth.

So what you, in theory, could do is to have a CQE without any field apart from an 'address' where the data of the c-record can be found. So when someone in Java calls SCQ.getEventId(); it will execute:
```
  long eventId = PlatformDependent.getLong(cqe + CQE_USER_DATA_FIELD);
```

This way you can still offer an idomatic Java API without having the extra data copy.

Reply:
Another approach:

```
public final class IOUringCqeAccessor {
 
    public static getEventId(long cqe) {
       return PlatformDependent.getLong(cqe + CQE_USER_DATA_FIELD);
    }

    public static int getRes(long cqe {
       return PlatformDependent.getInt(cqe + CQE_RES_FIELD);
    }

    public long getFlags(long cqe) {
       return  toUnsignedLong(PlatformDependent.getInt(cqe + CQE_FLAGS_FIELD));
    }
}
```

This way you also don't get litter and nu unwanted data copying.

Reply:
I would prefer the first approach 👍 

> So you have an intermediate copy of the data and this is one of the things io_uring tries to prevent if I understand the design correctly.

that's true, but these buffers for read/write operation are significantly larger (for example in netty 500 Bytes) than some integer/long variables

> Per entry, this would be 64 bytes of wasted memory bandwidth.

no the wasted stack memory bandwidth would be: int + long + long = 20 Bytes